api/service/version_service.go

```diff
@@ -84,25 +87,126 @@ func (service *versionsService) buildListVersionsQuery(modelID models.ID, query
+		}
 	}
 	return searchQuery
 }
 
+// getVersionSearchTerms parses a search query in this string format:
+//   <SEARCH_KEY_1>:<SEARCH_VAL_1> <SEARCH_KEY_2>:<SEARCH_VAL_2> ...
+// where pairs of search key-value are separated by whitespace, and returns
+// the search terms, a map of search key and the corresponding value.
+//
+// If the same search key is specified multiple times, the last one will override the former one.
+// Empty search key will be ignored since it has no meaning when searching.
+func getVersionSearchTerms(query string) map[string]string {
+	terms := map[string]string{}
+	tokens := strings.Split(query, ":")
+	key, val, nextKey := "", "" , ""
+	for i, token := range tokens {
+		token = strings.TrimSpace(token)
+		if i == 0 {
+			// First token is always the search key
+			key = token
+		} else if i == len(tokens) - 1 {
+			// Last token is always the search value
+			val = token
+		} else {
+			// Tokens in between contain both value for the current key and the next key
+			idx := strings.LastIndex(token, " ")
+			if idx >= 0 {
+				val = strings.TrimSpace(token[:idx])
+				nextKey = strings.TrimSpace(token[idx:])
+			} else {
+				// For cases like foo:bar:baz or foo::bar
+				val = ""
+				nextKey = token
+			}
+		}
+		if key != "" {
+			// Only consider non-empty keys because empty key has no meaning when searching
+			terms[key] = val
+		}
+		if i > 0 {
+			// For non-first token, it will contain value for the next key
+			key = nextKey
+		}
+	}
+	return terms
+}
+
+// versionLabelsRegex is the regular expression to match the labels key and values in the version
+// query by labels. The accepted characters in label keys and values roughly follows that in
+// Kubernetes labels:
+// https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names
+var versionLabelsRegex = regexp.MustCompile(`(?i)([a-z0-9A-Z-_]+)\s+in\s+\(([a-z0-9A-Z-_ ,]+)\),?\s*`)
+
+// generateLabelsWhereQuery parses the labels free text query in this format:
+//   <LABEL_KEY_1> IN (LABEL_VAL_1, LABEL_VAL_2,...), <LABEL_KEY_2> IN (LABEL_VAL1, LABEL_VAL2,...)
+// into a pair of SQL query template and arguments. The query template and arguments are
+// intended to be passed as arguments to go-orm db.Where method to filter rows by matching labels.
+//
+// In the free text query, label values for each key must be enclosed by a pair of parentheses.
+// Multiple label values for the same key are separated by "," character.
+//
+// For example, with the following labels query:
+//   animal in (cat, dog), color in (white)
+// generateLabelsWhereQuery will return query and args pairs which can be used to generate
+// this SQL statement:
+//   ( labels @> {"animal": "cat"} OR labels @> {"animal": "dog"} ) AND ( labels @> {"color": "white"} )
+// where the labels column is assumed to be in "jsonb" data type in Postgresql database.
+func generateLabelsWhereQuery(freeTextQuery string) (query string, args []interface{})  {
```

Comment:
I am wondering the query syntax that we want is uncommon that we need to write our own parser?
Is there a way to provide same semantics but with less custom part?

Reply:
yes this is described in of our proposal document for the search syntax and because of our use of jsonb data type.